### ISSUE TITLE
--exclude doesn't exclude when the pattern is an absolute path

### ISSUE THREAD

**Issue by adityasaky** [in-toto/in-toto#355]
Please fill in the fields below to submit an issue or feature request.  The
more information that is provided, the better.

**Description of issue or feature request**: `in-toto-run --exclude` doesn't actually exclude if the pattern provided is an absolute path. @lukpueh and I found this when trying to record artifacts from `/`.

**Current behavior**:
```
>>> from pathspec import PathSpec
>>> from in_toto.runlib import _apply_exclude_patterns
>>> exclude_patterns = ["/foo", "bar"]
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns or [])
>>> _apply_exclude_patterns(["/foo", "bar","baz"], exclude_filter)
['/foo', 'baz']
```
`/foo` appears after applying exclude patterns.

**Expected behavior**:
`/foo` should be removed.

**Comment by trishankatdatadog** [comment]
Good find!

**Comment by adityasaky** [comment]
Credit to @lukpueh, I was super confused.

**Comment by lukpueh** [comment]
Ping @kristelfung and @cindykimxp. Do I remember correctly that one of you volunteered to work on this?

**Comment by kristelfung** [comment]
> Ping @kristelfung and @cindykimxp. Do I remember correctly that one of you volunteered to work on this?

Hm, I don't think so!

**Comment by cindykimxp** [comment]
Ah, yeah I don't think I'm familiar with this one!

**Comment by lukpueh** [comment]
Thanks for getting back to me, and apologies for the false accusations. :P I must have mixed up the issues.

**Comment by adityasaky** [comment]
Relevant: https://github.com/cpburnz/python-path-specification/blob/master/pathspec/patterns/gitwildmatch.py

**Comment by benwuNYC** [comment]
The in-toto-run --exclude does not exclude patterns that include an absolute path (denoted by a forward slash ' / '). In the above example pinpointed by Aditya, this would be the '/foo' path. 

It was discovered that this issue has to do with the pathspec.patterns.gitwildmatch.py implementation of how patterns were converted into regular expressions.

**Link to pathspec.patterns.gitwildmatch.py implementation:** 

[https://github.com/cpburnz/python-path-specification/blob/master/pathspec/patterns/gitwildmatch.py#L88-L93
](url)

**Expected Behavior of pathspec.patterns.gitwildmatch.py**

In the above implementation, the pattern is split into segments (Line 84) with the forward slash as the separator, which removes the absolute path. For example, if `exclude_patterns = ["/foo"],` it would be split into `['',"foo"]`. Then, the segment would have the first empty segment removed (Line 93)  to make a pattern relative to the root. 

**Quick fix**

Including an extra forward slash in front of the absolute path solves the issue. For example, using `exclude_patterns = ["//foo", "bar"]` properly excludes the "/foo" absolute path after applying exclude patterns.

**Comment by adityasaky** [comment]
Great work, @itsbenwu!

> https://github.com/cpburnz/python-path-specification/blob/master/pathspec/patterns/gitwildmatch.py#L88-L93

I think it's important to note that the PathSpec implementation specifically intended this (I was quite surprised).

> Quick fix

This would fix it but as I mentioned yesterday, I'm not 100% sold on it. I think it'd also be interesting to see if this occurs on the equivalent pattern matching in in-toto-golang, and if not, we should aim to be consistent across them. Perhaps @shibumi can weigh in?

**Comment by lukpueh** [comment]
Thanks for digging into `gitwildmatch`, @itsbenwu and for your two cents, @adityasaky! I am okay with the *quickfix*, but we should document it.

To provide some background, when I first implemented artifacts excluding (in https://github.com/in-toto/in-toto/pull/56, worth a read), I used `fnmatch` for pattern matching and documented its behavior like so: 

https://github.com/in-toto/in-toto/blob/c0624605147b85cb1709355a5178a53dbee66f00/in_toto/runlib.py#L94-L115

`gitwildmatch` was added much later in https://github.com/in-toto/in-toto/pull/217 and the docstring was updated in https://github.com/in-toto/in-toto/pull/241. 

It would be nice if we could revise that docstring and add more comprehensive documentation. I picture a stand-alone document that describes the special characters and gives some examples, akin to the [settings documentation](https://in-toto.readthedocs.io/en/latest/command-line-tools/settings.html).  It's of course also okay to point to existing documentation about `gitwildmatch` and `.gitignore` as long as it is accurate and helpful.
We can then reference that document in the the relevant CLI docs, that is`--exclude` in `in_toto_{run, record}` and `exclude_patterns` in `in_toto_{run, record_start, record_stop}`. 

@itsbenwu, would you be willing to take a stab at this?

Regarding consistent behavior of exclude artifacts across in-toto(-python) and in-toto-golang, I think it's not a super high priority, because they are a non-specified convenience feature. That said, I would appreciate if @itsbenwu and @shibumi could coordinate on this. :)

**Comment by shibumi** [comment]
@lukpueh What about the special `**` operator? It's in every gitignore and not yet listed in your docstring for runlib.py. If we really support gitignore syntax we should add this operator. We definitely need to be precise if we do have full gitignore support or not. If not I do not  suggest to call it gitignore exclude syntax :)

Also I suggest having a look on the pattern format for gitignore: https://git-scm.com/docs/gitignore#_pattern_format

**Comment by lukpueh** [comment]
Thanks for chiming in, @shibumi! As indicated above, the documentation needs revision. And yes, if the pattern matcher we use supports `**`, then we should document it.

**Comment by shibumi** [comment]
ahh nevermind. I just realized I can scroll down and then see that you exclude the **-operator ^^

**Comment by benwuNYC** [comment]
Thanks for sharing some resources to help me out with the documentation, @lukpueh! I've created the documentation and followed the settings documentation as closely as possible, including documentation to Pathspec and .gitignore when appropriate. I shared a link to the documentation with you and it may also be accessed here: https://docs.google.com/document/d/1-99TyGUs0mSvFRUKpnQG2fffKYCNfQYbmE3VgPifBkY/edit?usp=sharing. 

When you get the chance, feel free to take a look at it and let me know if I should fix/edit anything!
